src/plugins/intel_cpu/src/nodes/gather.cpp

```diff
@@ -649,15 +650,58 @@ void Gather::execCompressed4Bit() {
 
                 OUT_TYPE* pdst = &dstData[dstIdx];
 
-                const size_t scale_offset = srcIdx / scale_group_size;
-                auto cur_zp = have_zp ? zp[srcIdx / zp_group_size] : 0;
                 size_t p = srcIdx;
                 size_t dst_idx = 0;
 
-                for (; p < srcIdx + afterAxisSize; p++) {
-                    auto val = srcData[p >> 1];
-                    pdst[dst_idx] = static_cast<OUT_TYPE>((get4Bit(val, p % 2) - cur_zp) * scale[scale_offset]);
-                    dst_idx++;
+                // heuristic:
+                // Case0: axis always equal to 0
+                // Case1: zp_group_size == scale_group_size
+                // Case2: have_1d_zp == true
+                // [Case0 + Case1] and [Case0 + Case2] take >99% probability
+                if (isAxisInputConst && axis == 0) {
+                    if (have_zp && (!have_1d_zp)) {
+                        // Case0 + Case1
+                        if (zp_group_size == scale_group_size) {
+                            for (; p < srcIdx + afterAxisSize; p+=scale_group_size) {
+                                const auto& cur_scale = scale[p / scale_group_size];
+                                const auto& cur_zp = zp[p / zp_group_size];
+                                for (size_t g = p; g < p + scale_group_size; g++) {
+                                    auto val = srcData[g >> 1];
+                                    pdst[dst_idx] = static_cast<OUT_TYPE>((get4Bit(val, g % 2) - cur_zp) * cur_scale);
+                                    dst_idx++;
+                                }
+                            }
```

Comment:
It seems to be too complicated to have 3 different optimized impls. Lets leave only that one and reference,
This one should be updated to cover case of `have_1d_zp`.  Non equal `scale_group_size` and `zp_group_size` (the case which I haven't seen in real models) can be computed via reference.

Reply:
Done